### PR TITLE
Reduce the number of badge requests

### LINKED ISSUE #396: Reduce the number of badge API requests sent
Ideas:

* Have a client-side blocklist of URL patterns or domains that we don't send badge requests for (current [server-side blocklist](https://github.com/hypothesis/h/blob/71cb4949ab8f850c9daa4c157362b562388d8140/h/views/badge.py#L28))
* Don't send the badge request for a newly-navigated URL until the user stays on that URL for a couple of seconds
* Have rate limiting for how quickly badge requests will be sent, no matter what
* Don't send requests if the URL for a tab changes from URL A to URL B where both URLs are the same as far as the backend is concerned. For example, if they only differ by fragment or scheme (HTTP vs HTTPS)

But really, just do whatever works. The aim is to significantly reduce the number of badge API requests being sent, while keeping the user experience good enough.

### PR DESCRIPTION
**Pull request by esanzgar**
It implements a custom debouncing mechanism.

A more standard approach like using `lodash._debounce` is not suited for
this particular case, because TabState is shared amoung all the tabs.

Reloading two tabs simulatneously shouldn't make one of the requests
to be cancelled.

The cancellation of a pending request could happen during the
waiting period or in-flight badge request.

Resolves #396

### CODE REVIEW

File: src/background/tab-state.js [hunk 1]
```diff
@@ -140,15 +143,47 @@ export default function TabState(initialState, onchange) {
...
-   * Request the current annotation count for the tab's URL
+   * Request the current annotation count for the tab's URL.
+   * It debounces previous requests within the value defined by
+   * WAITMILLISECONDS
    *
    * @method
    * @param {integer} tabId The id of the tab.
    * @param {string} tabUrl The URL of the tab.
    */
   this.updateAnnotationCount = function (tabId, tabUrl) {
```
Comment (lyzadanger) [inline src/background/tab-state.js:154]: The combination of `Promise`s and `async/await` here makes this method's flow a little hard to follow. Perhaps choose one?

One thing that would help would be to document this method's return type. e.g. it returns a `Promise` resolving to...?

File: src/background/tab-state.js [hunk 2]
```diff
@@ -140,15 +143,47 @@ export default function TabState(initialState, onchange) {
...
+      pendingRequest.cancel();
+    }
+
+    const fetchCount = async () => {
+      try {
+        const annotationCount = await uriInfo.getAnnotationCount(tabUrl);
+        this.setState(tabId, { annotationCount });
+      } finally {
+        // Clear the pending request
+        pendingAnnotationCountRequests.delete(tabId);
```
Comment (lyzadanger) [inline src/background/tab-state.js:169]: What happens if this method (`updateAnnotationCount`) is invoked again before this Promise settles (i.e. before the `pendingAnnotationCountRequests` map is updated? 

That is, what if the `timeout` has fired, this request is in-flight (and thus canceling the timeout won't do anything), and `updateAnnotationCount` is called?

File: src/background/tab-state.js [hunk 3]
```diff
@@ -140,15 +143,47 @@ export default function TabState(initialState, onchange) {
...
+        resolve(fetchCount());
+      }, WAITMILLISECONDS);
+
+      // Add the cancellation function that:
+      // 1. clears the timeout
+      // 2. rejects the promise
+      pendingAnnotationCountRequests.set(tabId, {
+        cancel: () => {
+          clearTimeout(timerId);
+          reject(new Error(`Canceled request of: ${tabUrl}`));
```
Comment (lyzadanger) [inline src/background/tab-state.js:184]: I'm wary of rejecting `Promise`s in what is a normal branch of code (i.e. not an error condition). It may seem a little counter-intuitive, but I'd recommend resolving to `noValue`/`null` and letting the caller deal with that value accordingly.

File: src/background/uri-info.js [hunk 4]
```diff
@@ -56,18 +60,19 @@ function query(uri) {
...
-      });
-  } else {
+
+  if (!shouldQueryUri(uri)) {
     return Promise.resolve(noValue);
   }
+
+  return query(uri)
+    .then(data => {
+      if (data || typeof data.total === 'number') {
```
Comment (lyzadanger) [inline src/background/uri-info.js:70]: Important! Needs to be `&&`

File: src/background/tab-state.js [hunk 5]
```diff
@@ -140,15 +143,47 @@ export default function TabState(initialState, onchange) {
...
+   * WAITMILLISECONDS
    *
    * @method
    * @param {integer} tabId The id of the tab.
    * @param {string} tabUrl The URL of the tab.
    */
   this.updateAnnotationCount = function (tabId, tabUrl) {
-    return uriInfo.getAnnotationCount(tabUrl).then(count => {
-      this.setState(tabId, { annotationCount: count });
+    const WAITMILLISECONDS = 3000; //
```
Comment (lyzadanger) [inline src/background/tab-state.js:155]: Might I suggest for readability `WAIT_MILLISECONDS`? Or you could abbreviate to `WAIT_MS`

File: src/background/uri-info.js [hunk 6]
```diff
@@ -1,26 +1,25 @@
...
-const BLOCKLIST = [
-  /* browser-specific URLs don't correspond to a meaningful document to annotate */
-  { urlProperty: 'protocol', value: 'chrome:' },
-  /* The following sites are personal in nature, high potential traffic
-     and URLs don't correspond to identifiable content */
-  { urlProperty: 'hostname', value: 'facebook.com' },
-  { urlProperty: 'hostname', value: 'www.facebook.com' },
-  { urlProperty: 'hostname', value: 'mail.google.com' },
-];
+const ALLOWEDPROTOCOLS = new Set(['http:', 'https:']);
```
Comment (lyzadanger) [inline src/background/uri-info.js:3]: Perhaps `ALLOWED_PROTOCOLS` (and `BLOCKED_HOSTNAMES` below)?

**Review by lyzadanger — COMMENTED** [review]
You general approach makes sense here and seems plausible—only allowing 0 or 1 debounced badge requests for a given tab and replacing a pending request if the tab is navigated. I'm having some challenges easily following the flow of `updateAnnotationCount` because of the mix of `async` and `Promise`s used. One thing I think would help is to document and explain the return type of `updateAnnotationCount` and what types the `Promise` could resolve to.

Perhaps choose one asynchronous approach—`Promise` _or_ `async/await`—and I can take another look?

**Review by esanzgar — COMMENTED** [review]
On this second iteration, I addressed the comments from @lyzadanger

File: src/background/tab-state.js (hunk 2 above)
Comment (robertknight) [reply src/background/tab-state.js:213]: In modern browsers [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) can be used to cancel `fetch` requests. If we wanted to use that API we could either bump the minimum Chrome version to one that supports it or just not implement cancellation in older versions of Chrome.

> That's why the cancelling function also rejects the pending promise.

I don't see any promise rejection happening in this code. The `cancel` callback clears the timeout and resolves it with a count of 0.

File: src/background/uri-info.js [hunk 7]
```diff
@@ -31,43 +30,51 @@ function shouldQueryUri(uri) {
...
+    }
+  );
+
+  const data = await response.json();
+
+  if (data && typeof data.total === 'number') {
+    return data.total;
+  }
+
+  throw new Error('Badge response has wrong format');
```
Comment (robertknight) [inline src/background/uri-info.js:65]: If parsing fails it doesn't necessarily mean that the format is "wrong". It might be that we change the API in future and an old extension is installed somewhere which is unable to parse it. "Unable to parse badge response" would be more accurate.

File: tests/background/tab-state-test.js [hunk 8]
```diff
@@ -132,23 +132,126 @@ describe('TabState', () => {
...
+      clock.tick(1010 + 2000);
+
+      await promise1;
+      assert.calledTwice(getStub); // request is not cancelled
+      assert.equal(tabState.getState(1).annotationCount, 0);
+      await promise2;
+      assert.equal(tabState.getState(1).annotationCount, testValue);
+    });
+
+    it('processes two consecutive request to the service if the requests are for different tabs', async () => {
```
Comment (robertknight) [inline tests/background/tab-state-test.js:225]: I know what you mean by this description, but I think perhaps the wording of this test description could do with an adjustment. I'm still trying to think of the best way to express this.

File: src/background/tab-state.js [hunk 9]
```diff
@@ -48,6 +48,12 @@ export default function TabState(initialState, onchange) {
   const self = this;
   let currentState;
 
+  // This variable contains a map between the tabId and a object with a
+  // cancellation function and the current waiting time.
+  // The cancellation function is used to abort the badge request in a debounce fashion
+  /** @type {Map<number, {cancel:function, waitMs: number}>} */
```
Comment (robertknight) [inline src/background/tab-state.js:54]: It could help here to introduce a `@typedef` which will allow you to document the `cancel` and `waitMs` properties more directly with `@prop`. The variable type can then be `/** @type {Map<number,BadgeRequest>} */` (where `BadgeRequest` is whatever you decide to call the type.

File: tests/background/tab-state-test.js [hunk 10]
```diff
@@ -132,23 +132,126 @@ describe('TabState', () => {
...
+      const testValue = 42;
+      getStub.resolves(testValue);
+      const tabState = new TabState({ 1: { state: states.ACTIVE } });
+
+      const promise1 = tabState.updateAnnotationCount(1, 'foobar.com');
+      clock.tick(800); // promise 1 is still waiting when promise2 is called
+      const promise2 = tabState.updateAnnotationCount(1, 'foobar.com');
+      clock.tick(2300);
+      await promise1;
+      assert.strictEqual(tabState.getState(1).annotationCount, 0);
```
Comment (robertknight) [inline tests/background/tab-state-test.js:175]: The code uses a mix of `assert.equal(...)` and `assert.strictEqual(...)`. We should be consistent about using one or the other in these tests. Either would be fine. We tend to use `assert.equal(...)` most of the time, although `assert.strictEqual(...)` can in principle catch certain additional mistakes.

**Review by robertknight — CHANGES_REQUESTED** [review]
I think we're getting close. Functionally the latest version of this code seems to work well and should reduce the number of badge requests significantly. The 1s initial delay with 3s maximum seems reasonable, though we can adjust this.

I had some suggestions regarding the implementation. The main one is that I found it tricky to follow the overall flow of events through the helpers and I think it would help to inline the logic, which is not that long, into the `updateAnnotationCount` method. I included a sketch of what this could look like in one of my comments for consideration. If we find that we need to do similar debouncing elsewhere in future then we can invent a suitable abstraction at that time.

A couple of other general things about the code:

- In async tests, try to avoid scenarios that are sensitive to the exact number of microtask ticks or turns of the event loop. These can be brittle and get broken by seemingly harmless refactoring. See my comments on the test that checks that the maximum wait time is 3000ms.
- Please try to avoid comments that repeat the code. These can get in the way of seeing the overall picture. Comments are good, but only when they "add value".

**Review by esanzgar — COMMENTED** [review]
@robertknight, thank you for your detailed comments and suggestions. They are really appreciated. I have tried to incorporate all of them.

File: src/background/tab-state.js [hunk 11]
```diff
@@ -86,6 +94,7 @@ export default function TabState(initialState, onchange) {
 
   this.clearTab = function (tabId) {
     this.setState(tabId, null);
+    pendingAnnotationCountRequests.delete(tabId);
```
Comment (robertknight) [inline src/background/tab-state.js:97]: This makes me realize that it would may make sense to store the pending request in the per-tab state. However this would require a bunch of other changes and refactoring so I suggest we should not try to do that in this PR.

**Review by robertknight — CHANGES_REQUESTED** [review]
Hi Eduardo,

I had another look through this. I have requests for some minor corrections. One point that I changed my thinking on is about `assert.equal(...)` vs `assert.strictEqual(...)`. I would like us to stick with `assert.equal(...)` except where there is a test case where the distinction is particularly important. This is mainly for consistency with the rest of our code, as it creates confusion otherwise if different modules have a different "default" without a clear reason.

File: src/background/uri-info.js [hunk 12]
```diff
@@ -31,43 +30,51 @@ function shouldQueryUri(uri) {
...
+
+  return true;
 }
 
 /**
  * Queries the Hypothesis service that provides
  * statistics about the annotations for a given URL.
+ *
+ * @return {Promise<number>}
+ * @throws Will throw a variety of errors: network, json parsing, or wrong format errors.
```
[the remark this answers is not in the document]
Comment (esanzgar) [reply src/background/uri-info.js:50]: `getAnnotationCount` catches all exception. I had a comment on the `catch` statement but I thought you requested to remove it. I am happy to add it back.

File: tests/background/tab-state-test.js [hunk 13]
```diff
@@ -34,15 +34,15 @@ describe('TabState', () => {
     it('replaces the current tab states with a new object', () => {
       state.load({ 2: { state: states.INACTIVE } });
       // `load` (re)sets all tabs to their default state, which is inactive
-      assert.equal(state.isTabActive(1), false);
-      assert.equal(state.isTabInactive(2), true);
+      assert.strictEqual(state.isTabActive(1), false);
+      assert.strictEqual(state.isTabInactive(2), true);
```
[the remark this answers is not in the document]
Comment (esanzgar) [reply tests/background/tab-state-test.js]: No problem, reverted.

File: src/background/tab-state.js (hunk 11 above)
Comment (esanzgar) [reply src/background/tab-state.js:97]: OK

**Review by esanzgar — COMMENTED** [review]
Let me know if I have addressed your concerns and suggestions.

File: src/background/uri-info.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (robertknight) [reply src/background/uri-info.js:50]: `getAnnotationCount` doesn't catch all exceptions on this branch! Can you have a look at it and check that it is doing what you intended.

File: src/background/tab-state.js [hunk 14]
```diff
@@ -48,6 +46,16 @@ export default function TabState(initialState, onchange) {
   const self = this;
   let currentState;
 
+  /**
+   * @typedef BadgeRequest - this object is used to cancel request in a debounce fashion
```
Comment (robertknight) [inline src/background/tab-state.js:50]:
```suggestion
   * @typedef BadgeRequest - this object is used to cancel a pending badge request
```

File: src/background/uri-info.js [hunk 15]
```diff
@@ -1,26 +1,26 @@
...
 
 /**
- * Should we send a "badge" request to obtain the annotation count for the
- * URL `uri`?
+ * Based on the request protocol and hostname decide if the URL should be sent to the "badge"
+ * request endpoint.
  *
  * @param {string} uri
- * @return {boolean}
+ * @return {boolean} - false if the URL should not be sent to the "badge" request
```
Comment (robertknight) [inline src/background/uri-info.js:23]:
```suggestion
 * @return {boolean} - false if the URL should not be sent to the "badge" request endpoint
```

**Review by robertknight — APPROVED** [review]
LGTM. Related Slack thread: https://hypothes-is.slack.com/archives/C4K6M7P5E/p1604318146454500?thread_ts=1603811687.273000&cid=C4K6M7P5E
